mantidproject/mantid
# DataProcessorUI: make pre-processing optional

Changes to `DataProcessorUI` to make pre-processing step optional:
- Previously, `GenericDataProcessorPresenter` (the widget's presenter) needed four arguments to its constructor: 
  1. A whitelist, defining the number of columns, their names and how they relate to the main reduction algorithm's properties
  2. A map containing pre-processing instructions (keys were columns that need pre-processing and values algorithms performing the pre-processing step)
  3. The main reduction algorithm
  4. A post-processing algorithm
- This PR is to make the pre-processing step optional. To resolve this I've used a [delegating constructor](https://github.com/mantidproject/mantid/blob/16631_Make_preprocessing_optional/MantidQt/MantidWidgets/src/DataProcessorUI/GenericDataProcessorPresenter.cpp#L106-L113) that, when no pre-processing map is given, just uses an empty one.

**To test:**
- Changes should not have any effect on the Reflectometry interface. Please, perform functional testing to ensure results are the same as before. The following reduction table (note that the instrument is wrong, INTER should be used instead of OFFSPEC), which I've just copied from https://github.com/mantidproject/mantid/pull/15943:

![intertest](https://cloud.githubusercontent.com/assets/8956985/15220007/cd3df0f6-185d-11e6-8627-3b8922c588b3.png)

should give the following results:
![intertestresults](https://cloud.githubusercontent.com/assets/8956985/15221360/04d84ca0-1863-11e6-8561-3c22611be721.png)
- There is just a minor change in the Reflectometry interface with respect to the master branch: the names of the generated workspaces. I have refactored the code to make workspace names more similar to what we had before:
  1. Loaded workspaces are called `TOF_12345` and `TRANS_12346` (assuming a run number `12345` and a transmission run `12346`)
  2. Reduced workspaces are called: `IvsQ_12345`, `IvsLam_12345` (for run number `12345`), instead of the previous `IvsQ_TOF_12345` and `IvsLam_TOF_12346`.
  3. Stitched workspaces are called: `IvsQ_12345_12345` (for run numbers `12345` and `12346`), instead of the previous `IvsQ_TOF_12345_TOF_12346`
- Ensure that the options `Plot` and `Plot Group` work as before
- Ensure that the generated notebook is correct

Fixes #16631.



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
